build-logic/src/main/kotlin/dokkabuild/tasks/GitCheckoutTask.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+package dokkabuild.tasks
+
+import org.eclipse.jgit.api.Git
+import org.gradle.api.DefaultTask
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.file.FileSystemOperations
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.TaskAction
+import javax.inject.Inject
+
+/**
+ * Clones a remote Git repository into [GitCheckoutTask.destination].
+ */
+@CacheableTask
+abstract class GitCheckoutTask @Inject constructor(
+    private val fs: FileSystemOperations
+) : DefaultTask() {
+
+    @get:OutputDirectory
+    abstract val destination: DirectoryProperty
+
+    /** Public HTTP URI of the Git repo. */
+    @get:Input
+    abstract val uri: Property<String>
+
+    /** Specific git hash to checkout. */
+    @get:Input
+    abstract val commitId: Property<String>
+
+    init {
+        group = "git checkout"
+    }
+
+    @TaskAction
+    fun action() {
+        val uri = uri.get()
+        val commitId = commitId.get()
+
+        fs.delete { delete(temporaryDir) }
+
+        Git.cloneRepository()
+            .setURI(uri)
+            .setDirectory(temporaryDir)
+            .call().use { git ->
+                git.pull().call()
+
+                git.checkout()
```

Comment:
is it intentional that we download full repository and then checkout specific commit instead of just downloading specific commit (head=1 or something like this)?

Reply:
That would definitely be an improvement. If you can figure out how to do that with JGit I would be grateful and impressed 😄

Reply:
Maybe `reset --hard $commitId` is all that's needed?

```kotlin
Git.cloneRepository()
    .setURI(uri)
    .setDirectory(temporaryDir)
    .call().use { git ->
        git.reset()
            .setMode(HARD)
            .setRef(commitId)
            .call()
    }
```

Reply:
maybe? :) 
```kotlin
        Git.cloneRepository()
            .setNoCheckout(true)
            .setURI(uri)
            .setDirectory(temporaryDir)
            .call().use { git ->
                git.checkout().setName(commitId).call()
            }
```

There is also `setDepth()` available on `cloneRepository` in 6.x, but not in 5.x version of jgit
Still, it should be fine any way, as it should still do this fast :)

Reply:
I discussed with @Tapchicoma and he showed me a similar process in `kotlin.git` - see https://github.com/JetBrains/kotlin/blob/f262247b2bcd7cc0e9d64d274f2fd412031187ab/libraries/tools/gradle/regression-benchmark-templates/src/main/kotlin/org/jetbrains/kotlin/gradle/benchmark/BenchmarkTemplate.kt#L133-L145